[BUG] val does not infer generic types in Java 14 switch expressions

**Describe the bug**
val fails to correctly infer generic type when used inside the switch expression block. It infers Object instead.

**To Reproduce**
```
import lombok.val;

import java.util.Map;

public class TestSwitchVal2 {

	public void test(int arg) {
		val x = switch (arg) {
			default -> {
				val map = Map.of("test", 1);
				yield map;
			}
		};
	}
}
```
C:\Users\boris\IdeaProjects\TestLombokJava14\src>javac -cp ..\lib\lombok-1.18.16.jar TestSwitchVal2.java
TestSwitchVal2.java:16: error: incompatible types: bad type in switch expression
                                yield map;
                                      ^
    Object cannot be converted to val
1 error

**Expected behavior**
map variable is expected to be inferred of correct type (Map<String,Integer).

**Version info (please complete the following information):**
 - Lombok version: 1.18.16
 - Platform: javac 14.0.2

Following code
```
public class TestSwitchVal1 {

	public void test(int arg) {
		val map = Map.of("test1", 1);
		val x = switch (arg) {
			default -> {
				val c = new HashMap<>(map);
				f(c);
				yield arg;
			}
		};
	}

	public void f(Map<String, Integer> m) {
	}
}
```
still does not compile with latest JDK 14 and 15.

```
TestSwitchVal1.java:21: error: method f in class TestSwitchVal1 cannot be applied to given types;
                                f(c);
                                ^
  required: Map<String,Integer>
  found:    Object
  reason: argument mismatch; Object cannot be converted to Map<String,Integer>
1 error
```

It does with JDK 16, though.

--- a/src/utils/lombok/javac/TreeMirrorMaker.java
+++ b/src/utils/lombok/javac/TreeMirrorMaker.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2010-2015 The Project Lombok Authors.
+ * Copyright (C) 2010-2021 The Project Lombok Authors.
  * 
  * Permission is hereby granted, free of charge, to any person obtaining a copy
  * of this software and associated documentation files (the "Software"), to deal
@@ -56,14 +56,14 @@ public class TreeMirrorMaker extends TreeCopier<Void> {
 	
 	@Override public <T extends JCTree> T copy(T original) {
 		T copy = super.copy(original);
-		originalToCopy.put(original, copy);
+		putIfAbsent(originalToCopy, original, copy);
 		return copy;
 	}
 	
 	@Override public <T extends JCTree> T copy(T original, Void p) {
 		T copy = super.copy(original, p);
-		originalToCopy.put(original, copy);
-		return copy;
+		putIfAbsent(originalToCopy, original, copy);
+		return copy; 
 	}
 	
 	@Override public <T extends JCTree> List<T> copy(List<T> originals) {
@@ -71,7 +71,7 @@ public class TreeMirrorMaker extends TreeCopier<Void> {
 		if (originals != null) {
 			Iterator<T> it1 = originals.iterator();
 			Iterator<T> it2 = copies.iterator();
-			while (it1.hasNext()) originalToCopy.put(it1.next(), it2.next());
+			while (it1.hasNext()) putIfAbsent(originalToCopy, it1.next(), it2.next());
 		}
 		return copies;
 	}
@@ -81,7 +81,7 @@ public class TreeMirrorMaker extends TreeCopier<Void> {
 		if (originals != null) {
 			Iterator<T> it1 = originals.iterator();
 			Iterator<T> it2 = copies.iterator();
-			while (it1.hasNext()) originalToCopy.put(it1.next(), it2.next());
+			while (it1.hasNext()) putIfAbsent(originalToCopy, it1.next(), it2.next());
 		}
 		return copies;
 	}
@@ -119,5 +119,11 @@ public class TreeMirrorMaker extends TreeCopier<Void> {
 	// This and visitVariable is rather hacky but we're working around evident bugs or at least inconsistencies in javac.
 	@Override public JCTree visitLabeledStatement(LabeledStatementTree node, Void p) {
 		return node.getStatement().accept(this, p);
+	}
+	
+	private <K, V> void putIfAbsent(Map<K, V> map, K key, V value) {
+		if (!map.containsKey(key)) {
+			map.put(key, value);
+		}
 	}
 }

--- a/test/transform/resource/after-delombok/ValSwitchExpression.java
+++ b/test/transform/resource/after-delombok/ValSwitchExpression.java
@@ -1,0 +1,11 @@
+// version 14:
+public class ValSwitchExpression {
+	public void method(int arg) {
+		final int x = switch (arg) {
+			default -> {
+				final java.lang.String s = "string";
+				yield arg;
+			}
+		};
+	}
+}

--- a/test/transform/resource/after-ecj/ValSwitchExpression.java
+++ b/test/transform/resource/after-ecj/ValSwitchExpression.java
@@ -1,0 +1,16 @@
+// version 14:
+import lombok.val;
+public class ValSwitchExpression {
+  public ValSwitchExpression() {
+    super();
+  }
+  public void method(int arg) {
+    final @val int x =     switch (arg) {
+    default ->
+        {
+          final @val java.lang.String s = "string";
+          yield arg;
+        }
+    };
+  }
+}

--- a/test/transform/resource/before/ValSwitchExpression.java
+++ b/test/transform/resource/before/ValSwitchExpression.java
@@ -1,0 +1,13 @@
+// version 14:
+import lombok.val;
+
+public class ValSwitchExpression {
+	public void method(int arg) {
+		val x = switch (arg) {
+			default -> {
+				val s = "string";
+				yield arg;
+			}
+		};
+	}
+}

